### PR TITLE
API First and Agility

### PR DESCRIPTION
**Pull request by tfrauenstein**
Elaborate API First Principle and its relation to agile practices

### CODE REVIEW

File: general-guidelines/GeneralGuidelines.md [hunk 1]
```diff
@@ -2,15 +2,60 @@
...
+
+- single source of truth for API specification; 
+  it is a crucial part of a contract between service provider and client users
+- infrastructure tooling for API discovery, API GUIs, API documents, automated quality checks
+
+An element of API First are also this API Guidelines and a [lightweight API review process \[internal link\]](https://github.bus.zalan.do/ApiGuild/ApiReviewProcedure) as to get early review feedback from peers and client developers. 
+Peer review is important for us to get high quality APIs, to enable architectural and design alignment 
+and to supported development of client applications decoupled from service provider engineering life cycle. 
+
+It is important to learn, that API First principle is **not in conflict with the agile development principles** that we love.
```
Comment (ePaul) [inline general-guidelines/GeneralGuidelines.md]: "that API first principle" → "that the API First principle" or "that our API First principle"

File: general-guidelines/GeneralGuidelines.md [hunk 2]
```diff
@@ -2,15 +2,60 @@
...
-Whether you choose YAML or JSON as a format of your OpenAPI API specification file is up to you; 
+We use [OpenAPI specification](http://swagger.io/specification/) (aka Swagger spec) as standard for our REST API definitions. 
+You may choose YAML or JSON as a format of your OpenAPI API specification file; 
 however, YAML is generally preferred due to its improved readability. 
+
 We also call the OpenAPI API specification the "API Reference Definition"; 
-it provides all information needed by an experienced API client developer to use this API. 
+it provides all information needed by an experienced API client developer to use this API.
 
+The OpenAPi API specification file should be subject of version control together with source code management.
```
Comment (ePaul) [inline general-guidelines/GeneralGuidelines.md]: "OpenAPi" → "OpenAPI"

File: general-guidelines/GeneralGuidelines.md [hunk 3]
```diff
@@ -2,15 +2,60 @@
...
+You may choose YAML or JSON as a format of your OpenAPI API specification file; 
 however, YAML is generally preferred due to its improved readability. 
+
 We also call the OpenAPI API specification the "API Reference Definition"; 
-it provides all information needed by an experienced API client developer to use this API. 
+it provides all information needed by an experienced API client developer to use this API.
 
+The OpenAPi API specification file should be subject of version control together with source code management. 
+Service also have to provide 
+[online access to the API Reference definition](../api-operation/ApiOperation.md#must-Provide-Online-Access-to-OpenAPI-Reference-Definition) for its external API(s).
```
Comment (ePaul) [inline general-guidelines/GeneralGuidelines.md]: "Service" → "Services" + "its" → "their"

or 

"Service also have" → "A Service also has".

File: general-guidelines/GeneralGuidelines.md [hunk 4]
```diff
@@ -2,15 +2,60 @@
...
+API may change and profit from implementation concerns and automated testing feedback. 
+API evolution during development life cycle may include breaking changes for not yet productive features 
+and as long as we have aligned the changes with the clients. 
+Hence, API First does *not* mean that you must have 100% domain and requirement understanding and can never produce code 
+before you have defined the complete API and get it confirmed by peer review. On the other hand, API First obviously is 
+in conflict with the practise of publishing API definition and asking for peer review after the service integration 
+or even the service productive operation has started. 
+It is crucial to request and get early feedback — as early as possible, unless the API changes are focussed 
+to the next evolution step and have a certain quality (including API Guideline compliance), 
+already confirmed by team internal reviews.
```
Comment (ePaul) [inline general-guidelines/GeneralGuidelines.md]: Hmm, not sure what the part after the "–" is meant to mean. Evolutionary API changes don't need early review?

File: general-guidelines/GeneralGuidelines.md [hunk 5]
```diff
@@ -2,15 +2,60 @@
...
+Service applications should evolve incrementally — and so its APIs. Of course, API specification will 
+and should evolve iteratively in different cycles, each starting with draft status and early team 
+and peer review feedback. 
+
+API may change and profit from implementation concerns and automated testing feedback. 
+API evolution during development life cycle may include breaking changes for not yet productive features 
+and as long as we have aligned the changes with the clients. 
+Hence, API First does *not* mean that you must have 100% domain and requirement understanding and can never produce code 
+before you have defined the complete API and get it confirmed by peer review. On the other hand, API First obviously is 
+in conflict with the practise of publishing API definition and asking for peer review after the service integration
```
Comment (ePaul) [inline general-guidelines/GeneralGuidelines.md]: "practise" → "practice" (the former is just an british form of the verb, the substantive is always with "c".)

File: general-guidelines/GeneralGuidelines.md [hunk 6]
```diff
@@ -2,15 +2,60 @@
...
+- single source of truth for API specification; 
+  it is a crucial part of a contract between service provider and client users
+- infrastructure tooling for API discovery, API GUIs, API documents, automated quality checks
+
+An element of API First are also this API Guidelines and a [lightweight API review process \[internal link\]](https://github.bus.zalan.do/ApiGuild/ApiReviewProcedure) as to get early review feedback from peers and client developers. 
+Peer review is important for us to get high quality APIs, to enable architectural and design alignment 
+and to supported development of client applications decoupled from service provider engineering life cycle. 
+
+It is important to learn, that API First principle is **not in conflict with the agile development principles** that we love. 
+Service applications should evolve incrementally — and so its APIs. Of course, API specification will
```
Comment (ePaul) [inline general-guidelines/GeneralGuidelines.md]: "API specification" needs either an article, or a plural. (Same in the next some sentences.)

File: general-guidelines/GeneralGuidelines.md [hunk 7]
```diff
@@ -2,15 +2,60 @@
...
+already confirmed by team internal reviews. 
+
+
 ## {{ book.must }} Provide API Reference Definition using OpenAPI
 
-API First is one of our architecture principles, as per our Architecture Rules of Play.
-Define, document and review your APIs before delivering them (also see API review procedure).
-Whether you choose YAML or JSON as a format of your OpenAPI API specification file is up to you; 
+We use [OpenAPI specification](http://swagger.io/specification/) (aka Swagger spec) as standard for our REST API definitions. 
+You may choose YAML or JSON as a format of your OpenAPI API specification file;
```
Comment (ePaul) [inline general-guidelines/GeneralGuidelines.md]: About wording: I would use "OpenAPI specification" just for the specification document (the one linked there). The individual files (aka *.yaml/*.json) I would call "OpenAPI definition".

I'm not sure how common my approach is, though. Opinions?

File: general-guidelines/GeneralGuidelines.md [hunk 8]
```diff
@@ -2,15 +2,60 @@
...
+
+- define APIs outside the code first using a standard specification language
+- get early review feedback from peers and client developers
+
+By defining APIs outside the code, we want to facilitate early review feedback and also a development 
+discipline that focus service interface design on...
+
+- profound understanding of the domain and required functionality
+- generalized business entities / resources — avoidance of use case specific APIs
+- clear separation of WHAT vs. HOW concerns, i.e. abstraction from implementation aspects
```
Comment (ePaul) [inline general-guidelines/GeneralGuidelines.md]: Not sure if this is included here, or whether we should mention it separately: It also makes it easier to replace the implementation (even using a different programming language/framework) without changing the API.

Comment (tfrauenstein) [inline general-guidelines/GeneralGuidelines.md]: it is included, but I added it to emphasis

File: general-guidelines/GeneralGuidelines.md (hunk 1 above)
Comment (tfrauenstein) [inline general-guidelines/GeneralGuidelines.md]: done

File: general-guidelines/GeneralGuidelines.md (hunk 6 above)
Comment (tfrauenstein) [inline general-guidelines/GeneralGuidelines.md]: done

File: general-guidelines/GeneralGuidelines.md (hunk 5 above)
Comment (tfrauenstein) [inline general-guidelines/GeneralGuidelines.md]: great English hints :-) -- done

File: general-guidelines/GeneralGuidelines.md (hunk 4 above)
Comment (tfrauenstein) [inline general-guidelines/GeneralGuidelines.md]: of course, early feedback is need for evolutionary API changes -- should ne clearer now

File: general-guidelines/GeneralGuidelines.md (hunk 7 above)
Comment (tfrauenstein) [inline general-guidelines/GeneralGuidelines.md]: I generally use the same wording -- correction done.

File: general-guidelines/GeneralGuidelines.md (hunk 2 above)
Comment (tfrauenstein) [inline general-guidelines/GeneralGuidelines.md]: done

File: general-guidelines/GeneralGuidelines.md (hunk 3 above)
Comment (tfrauenstein) [inline general-guidelines/GeneralGuidelines.md]: done

**Comment by ePaul** [comment]
:+1:

**Comment by codekultur** [comment]
👍

**Pull request merged**